Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs

```diff
@@ -686,17 +695,8 @@ protected static void PropagateBrakeLinePressures(float elapsedClockSeconds, Tra
                             // Appears to work on the principle of pressure equalisation, ie trailing car pressure will drop by fraction of total volume determined by lead car.
 
                             // TODO - Confirm logic - it appears to allow for air flow to coupled car by reducing (increasing) preceeding car brake pipe - this allows time for the train to "recharge".
-                            if (car0 == lead) // If this is the locomotive
-                            {
-                                car.BrakeSystem.BrakeLine1PressurePSI -= TrainPipePressureDiffPropogationPSI * brakePipeVolumeM30 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
-                                car0.BrakeSystem.BrakeLine1PressurePSI += TrainPipePressureDiffPropogationPSI * car.BrakeSystem.BrakePipeVolumeM3 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
-                                car0.BrakeSystem.BrakeLine1PressurePSI = MathHelper.Clamp(car0.BrakeSystem.BrakeLine1PressurePSI, 0.001f, lead.TrainBrakeController.MaxPressurePSI); // Prevent Lead locomotive pipe pressure increasing above the value calculated above
-                            }
-                           else
-                            {
-                                car.BrakeSystem.BrakeLine1PressurePSI -= TrainPipePressureDiffPropogationPSI * brakePipeVolumeM30 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
-                                car0.BrakeSystem.BrakeLine1PressurePSI += TrainPipePressureDiffPropogationPSI * car.BrakeSystem.BrakePipeVolumeM3 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
-                            }                      
+                            car.BrakeSystem.BrakeLine1PressurePSI -= TrainPipePressureDiffPropogationPSI * brakePipeVolumeM30 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
+                            car0.BrakeSystem.BrakeLine1PressurePSI += TrainPipePressureDiffPropogationPSI * car.BrakeSystem.BrakePipeVolumeM3 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
```

Comment:
From memory, as the pressure change is propagated along the train pipe a certain amount of change happens in each of the cars (as suggested by + and - statements). If the locomotive pressure is not clamped, then its brake pipe pressure can exceed the maximum pressure (this maybe what you are trying to achieve in your change, but in a normal system it may not be the case).

Changes to this section of code may impact on the propagation delay along the train pipe, so any changes should be tested as a before and after to make sure that timings have not varied.

Reply:
I haven't found any difference in brake pipe charging and discharging times. Note that when applying and releasing brakes, brake pipe pressure will be below maximum pressure, so the clamp will have no effect. Also, TrainPipePressureDiffPropagationPSI can't be greater than the difference of pressures, so the pressures won't increase magically.

However, I have seen something which might cause problems:
`float TrainPipePressureDiffPropogationPSI = Math.Min (TrainPipeTimeVariationS * (p1 - p0) / brakePipeTimeFactorS, p1 - p0);`
This is OK when p1 > p0, but when p1 < p0 I don't think this code is doing what the writer expected. What's your opinion about it?

Reply:
> "This is OK when p1 > p0, but when p1 < p0 I don't think this code is doing what the writer expected. What's your opinion about it?"

I am sorry but I don't understand the issue, as I have always interpreted it as creating either a -ve or +ve number depending upon whether the train pipe is charging or releasing. There appears to be no separate propagation code for releasing/charging, and hence the one code section needs to handle both. The Min statement is just taking minimum +ve or -ve number to use as the difference value.

> "Note that when applying and releasing brakes, brake pipe pressure will be below maximum pressure, so the clamp will have no effect."

Is this the case, the way that I have interpreted the code, is as follows.

When the brakes are fully released, the BP = MaxPressure, lets say 70psi. So for the locomotive (car0), OR process the following line of code,

`car0.BrakeSystem.BrakeLine1PressurePSI += TrainPipePressureDiffPropogationPSI * car.BrakeSystem.BrakePipeVolumeM3 / (brakePipeVolumeM30 + car.BrakeSystem.BrakePipeVolumeM3);
`
So in this instance wouldn't we get the following result, BP = BP + Diff = 70+ psi? The clamp statement would then reduce the loco value to 70psi? Hence it is providing a form of feedback for the locomotive pressure.

Other cars in the train are treated differently, ie only the subtraction statement is used. (Either it goes up or down depending upon whether Diff is -ve or +ve.)

Unfortunately, at the moment I haven't got time to get my head fully back around this code, so I am not able to fully document any specific impacts, apart from the "vague" memory and comments already made for these proposed changes. Suffice to say that like a lot of other code in OR, it is quite complex, and often very hard to understand the logic used for its creation by the original developer. Typically I have found it better to leave code alone unless I can fully understand the logic for its original creation.

Reply:
Consider the situation of brake release. Brake pipe pressure in car0 (lead locomotive) will be 70 psi, and pressure in another car will be lower than 70, e. g., 60 psi. Then, p1-p0 = -10 psi. This illustrates the two points I'm trying to tell:
- First: `p1-p0` is -10, but `TrainPipeTimeVariationS * (p1 - p0) / brakePipeTimeFactorS` will be around -9 psi. The minimum of both magnitudes is -10. However, if both were positive, the minimum would be 9. I'm pretty sure that what the original author meant was: 
`float TrainPipePressureDiffPropogationPSI = (p1>p0 ? 1 : -1) * Math.Min(TrainPipeTimeVariationS * Math.Abs(p1 - p0) / brakePipeTimeFactorS, Math.Abs(p1 - p0));`
The way it is written now it's not symmetric, but one would expect symmetry from a physical system (the brake pipe) which has no preferred direction.

- Second: p1-p0 < 0, so TrainPipePressureDiffPropagationPSI < 0. Therefore, pressure in car0 will be decreased, and increased in car1. The algorithm used here works by the principle of pressure equalization between adjacent cars, so air will flow from the car with higher pressure to the car with lower, never the other way round.

Reply:
> "The way it is written now it's not symmetric, but one would expect symmetry from a physical system (the brake pipe) which has no preferred direction."

I would agree with your thoughts in regard to the symmetry, and would be happy to see the Diff value changed accordingly. It should have a developer comment describing the intent of the action for future developers. 

I don't think that the suggested change will work. It may also require an ABS function around the first factor as well.

> "Second: p1-p0 < 0, so TrainPipePressureDiffPropagationPSI < 0."

If p0 = 60, and p1 = 70, ie brake pipe needs to reduce pressure along the train. In this case p1-p0 > 0, so TrainPipePressureDiffPropagationPSI > 0. Therefore, pressure in car0 will be increased, and decreased in car1. Whilst the calculation of the Diff value and the Volume offsetting may mean it might not go above the MaxPressure, it will probably still be good to retain the Clamp statement as a safeguard.

Reply:
The clamp is what prevents the brake pressure from being overcharged, which is what I'm trying to achieve. I have studied the formula and I'm sure that pressure won't increase over the maximum in other situations. If that happens, it will be in another part of the code, and it should be identified and fixed there.
There are some vehicles where this clamp has sense, as they are overcharge-protected, but this should be set on a per car basis, and not just for the first car in the train.

Reply:
> "I have studied the formula and I'm sure that pressure won't increase over the maximum in other situations. If that happens, it will be in another part of the code, and it should be identified and fixed there."

    As long as the current overcharge-protected functionality is preserved, then it is fine for the code changes. However it would be good to validate the fact that it does/doesn't need to be corrected elsewhere in the code through some form of testing, etc, rather then "breaking" the code, and hoping that somebody notices it.

> "There are some vehicles where this clamp has sense, as they are overcharge-protected, but this should be set on a per car basis, and not just for the first car in the train."

I am sorry but I don't understand this.

Given the fact that we are talking about clamping the brake pipe at a max pressure, and that the locomotive is the only car in the train that controls changes in the brake pipe (apart from minor leakage), why would it be necessary to set it on a per car basis? The locomotive is the overall pressure reference for the brake pipe on the whole train, and the pressure in all other cars should follow accordingly.

Reply:
> As long as the current overcharge-protected functionality is preserved, then it is fine for the code changes. However it would be good to validate the fact that it does/doesn't need to be corrected elsewhere in the code through some form of testing, etc, rather then "breaking" the code, and hoping that somebody notices it.

I have checked all formulas and pressure will never increase over the maximum without overcharge feature. However pressure can become negative, so I have added some checks.

> Given the fact that we are talking about clamping the brake pipe at a max pressure, and that the locomotive is the only car in the train that controls changes in the brake pipe (apart from minor leakage), why would it be necessary to set it on a per car basis? The locomotive is the overall pressure reference for the brake pipe on the whole train, and the pressure in all other cars should follow accordingly.

In my opinion, if overcharge protection (i. e. pressure clamp) were implemented it should be done depending on each car. There may be the case where the locomotive tries to overcharge the brake pipe but some of the wagons are overcharge protected. The pressure should be clamped when at least one vehicle has this protection. Vehicles with overcharge protection are not a common situation, though.

Reply:
My understanding of a non-overcharge triple valve is it just follows the train brake pipe pressure. In railway operating instructions I have seen, it is suggested that the train brake pipe pressure can be varied by the crew depending upon the type of train that it is. Hence the triple valve works at these different pressures without any individual alterations, and in this instance the  brake pipe pressure sets the maximum operating pressure of the triple valve, and thus would only need to be clamped on the locomotive. 

At the moment I can't speak for the (overcharge) triple valves that you have described. Do you have any documents that you can point me to?

Reply:
I don't think there exist such 'overcharge triple valves'. Overcharge is a situation that might happen, and all triple valves will react to it in a way or another. Some of them will have overcharge protections, so as soon as the BP pressure is greater (e. g. 5.3 bar) than aux pressure (say 5.1 bar) the valve will release the brakes but the auxiliary pressure won't be increased. Other valves (generally older ones) will simply equalize both pressures. In both situations, if the auxiliary pressure is higher than BP's, it should slowly decrease while the valve is in release position, so in the end auxiliary pressure is equal to normal system pressure (this is what I have introduced with brake insensitivity parameter, see https://rail21.pagesperso-orange.fr/Freinage_Distributeur_EN.pdf for the timings). If auxiliary reservoir pressure weren't decreased, then brakes would be applied even if the controller was in release position. This is what happens if overcharge elimination process is interrupted, leading to brakes being wrongly applied. Some systems (https://patents.google.com/patent/US6375275B1/en) will act if overcharge is not correctly eliminated.

There are two types of overcharge:
 - Undesired overcharge: usually happens when the brake pipe is connected for a long time to the main reservoir in order to achieve a fast brake pipe charging.
 - Intentional overcharge: the driver can overcharge the brake pipe on purpose to release stuck brakes, especially in twin pipe systems. For example, if a locomotive with max pressure of 5.2 bar is replaced with a 4.9 bar one, the control reservoirs in brake distributors remain at 5.2 bar, so the new locomotive can't fully release all brakes. The driver momentarily increases pressure in BP over 5.3 bar, so the brakes are released. Then, the pressure starts decreasing very slowly, and the control reservoir pressure is finally 4.9 bar, so the new locomotive can operate the brakes normally. Something similar happens in single pipe systems.

Reply:
Thanks for the comments.

If I understand what you are saying, the triple valve needs to be treated differently depending upon which type it is:

i) Older Style - will only respond to the train brake pipe. In this type the pressure should decrease at the "current" release rate, as it is merely following the train brake pipe pressure (as it has always done).

ii) "Overcharge" style - will behave as you have described, and decrease at the rates that you have indicated.

So perhaps the only difference between the two will be the decrease rate value.

Reply:
Current implementation causes auxiliary reservoir pressure to not decrease at all if the valve is in Release position. If it decreased following BP pressure, brakes would never be applied. Therefore, the flow from auxiliary reservoir to brake pipe, if any, has to be slow.

I don't think it's necessary to have two different types of triple valves. If you have never found any triple valve with overcharge protections, they must be really uncommon, if they actually exist.

The thing here is, how do _normal_ triple valves react to an overcharge? They aren't able to differentiate overcharge from normal releases, as they have no control reservoir to check against. Therefore, they will charge auxiliary reservoir with a pressure equal to brake pipe's. Unless I'm missing some weird exception, this happens with all triple valves.

Another different question is: how do triple valves react to a really slow decrease of BP pressure, while in Release position? This situation happens both when overcharge is eliminated and due to small leakages. When the valve is in Lap position, the answer is simply that brakes are applied until BP and aux have the same pressure again. However, if the valve is in Release position, the UIC allowing a 'brake pipe insensitivity' suggests that brakes aren't applied if the pressure reduction is small.

Distributors know when the brake pipe is overcharged, as BP pressure will be higher than control reservoir pressure. Some of them might react sealing the reservoirs so only the BP is overcharged (this is the only example of overcharge-protected equipment). Currently OR uses emergency reservoir as a fixed pressure control reservoir, so it won't be affected by overcharge. However, the behaviour of the control reservoir in OR seems to have some differences with what I have read in manuals.